### PR TITLE
WishesAPI Fixes

### PR DESCRIPTION
**Pull request by jinyingtan**
In this PR:
- Changes the implementation for getting a specific item from firebase
- Update docs for WishesAPI
- For functions that return boolean, it throws errors and returns the object
- Minor Query Optimization

Query Optimization Results 
- Approach: Used `console.time()` at function start and `console.timeEnd()` before function return
- `createWish`
  - Before: 1051ms
  - After: 760ms
- `updateWish`
  - Before: 1285ms, 1135ms, 1284ms
  - After: 979ms, 1189ms, 1085ms
- `completeWish`
  - Before: 673ms, 675ms, 1035ms
  - After: 647ms, 540ms, 544ms

### CODE REVIEW

File: utils/api/wishes.js [hunk 1]
```diff
@@ -85,12 +86,49 @@ class WishesAPI {
   }
 
   /**
-   * Get the initial batch of pending wishes belonging to a category. Only return results of WISHES_BATCH_SIZE
-   * @param {object} categoryId The category id
+   * Gets a batch of all pending wishes. Only return results of WISHES_BATCH_SIZE
+   * @param {string} orderBy The way to order the wishes
```
Comment (kohchihao) [inline utils/api/wishes.js:90]: What is `The way to order the wishes` ??? What are the other ways?

File: utils/api/wishes.js [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 
 class WishesAPI {
   /**
    * Create a new wish
-   * @param {string} title The wish title
-   * @param {string} description The wish description
-   * @param {array} categories A list of categories that the wish belongs to
+   * @param {string} title The wish title text
+   * @param {string} description The wish description text
+   * @param {array} categories A list of categories names that the wish belongs to
```
Comment (kohchihao) [inline utils/api/wishes.js:14]: Should we standardise if your API takes in `category-name` or `category-id` ??

File: utils/api/wishes.js [hunk 3]
```diff
@@ -85,12 +86,49 @@ class WishesAPI {
...
+  /**
+   * Get a batch of pending wishes belonging to a category. Only return results of WISHES_BATCH_SIZE
+   * @param {string} categoryId The category id
    * @param {string} orderBy The way to order the wishes. Only wishesSortTypeAllowed
    * @param {boolean} isReverse Indicates if the query should be ordered in reverse
    * @param {object} lastQueriedDocument The last queried firebase document to start the query after. If the field is not given, the query will start from the first document
-   * @throws {ReferenceError}
+   * @throws {WishError}
    * @throws {FirebaseError}
    * @return {object} A firebase document of ordered pending wishes belonging to a category
```
Comment (kohchihao) [inline utils/api/wishes.js:133]: Should the return type be `array` or `object` ?

File: utils/api/wishes.js [hunk 4]
```diff
@@ -141,42 +175,42 @@ class WishesAPI {
...
    * Get the initial batch of wishes belonging to a NPO filter by its status. Only return results of WISHES_BATCH_SIZE
    * @param {string} npoId
-   * @param {string} status The status of the wishes
+   * @param {string} status The status of the wishes to query
    * @param {object} lastQueriedDocument The last queried firebase document to start the query after. If the field is not given, the query will start from the first document
    * @throws {FirebaseError}
-   * @return {object} A firebase document of wishes belonging to a NPO, filtered by status
+   * @return {array} A list of firebase document of wishes belonging to a NPO, filtered by status
    */
   async getNPOWishesFilterByStatus(npoId, status, lastQueriedDocument = null) {
```
Comment (kohchihao) [inline utils/api/wishes.js:215]: Is it possible to combine this `getNPOWishesFilterByStatus` with `getNPOWishes` ??? Maybe make `status` optional for `getNPOWishes`.

Comment (kohchihao) [reply utils/api/wishes.js:215]: The 2 API look rather similar.

File: utils/api/wishes.js (hunk 1 above)
Comment (jinyingtan) [reply utils/api/wishes.js:90]: There are several ways to order the wishes?

Comment (kohchihao) [reply utils/api/wishes.js:90]: How would the other dev know what are the other ways. Maybe can write `The way to order the wishes. Can find other ways in xxxxx.js`

File: utils/api/wishes.js (hunk 4 above)
Comment (jinyingtan) [reply utils/api/wishes.js:215]: It is possible to just combine to one, but would still need use to different queries which are currently used in both the function. I can extract it to provide one interface but still have to handle the query separately. Let me know if this works

Comment (jamessspanggg) [reply utils/api/wishes.js:204]: > I can extract it to provide one interface but still have to handle the query separately.

Sounds ok to me

**Comment by jamessspanggg** [comment]
@kohchihao can take a look at this?

File: utils/api/wishes.js [hunk 5]
```diff
@@ -1,48 +1,43 @@
...
+    // time: 1051ms -> 760ms
     let userInfo = {};
     let organizationInfo = {};
 
-    // Used a for loop instead of a forEach because forEach is not promise aware
-    for (let i = 0; i < categories.length; i++) {
-      let categoryInfo = await this._getCategoryInfoByName(categories[i]);
+    const categoriesInfoPromise = this._getAllCategoriesInfo(categories);
+    const allUserInfoPromise = this._getCurrentUserInfo();
+    const [categoriesInfo, allUserInfo] = await Promise.all([categoriesInfoPromise, allUserInfoPromise]);
```
Comment (kohchihao) [inline utils/api/wishes.js:26]: Looks good!

File: utils/api/wishes.js [hunk 6]
```diff
@@ -85,51 +80,79 @@ class WishesAPI {
...
+   * @returns {array} A list of firebase document of all ordered pending wishes
    */
-  async getPendingWishesForCategory(
-    categoryId,
-    orderBy = WishesSortTypeConstant.TIMESTAMP,
-    isReverse = false,
-    lastQueriedDocument = null
-  ) {
+
+  async getAllPendingWishes(orderBy = TIMESTAMP, isReverse = false, lastQueriedDocument = null) {
```
[the remark this answers is not in the document]
Comment (kohchihao) [reply utils/api/wishes.js:92]: agreed. `getPendingWishes` is good enough.

File: utils/api/wishes.js [hunk 7]
```diff
@@ -85,51 +80,79 @@ class WishesAPI {
...
   /**
-   * Get the initial batch of pending wishes belonging to a category. Only return results of WISHES_BATCH_SIZE
-   * @param {object} categoryId The category id
-   * @param {string} orderBy The way to order the wishes. Only wishesSortTypeAllowed
+   * Gets a batch of all pending wishes. Only return results of WISHES_BATCH_SIZE
+   * @param {string} orderBy The way to order the wishes. Look at wishesSortType.js to know what are the various ways.
    * @param {boolean} isReverse Indicates if the query should be ordered in reverse
    * @param {object} lastQueriedDocument The last queried firebase document to start the query after. If the field is not given, the query will start from the first document
-   * @throws {ReferenceError}
+   * @throws {DonationError}
```
Comment (kohchihao) [inline utils/api/wishes.js:87]: Should `DonationError` be `WishError` ?

File: utils/api/wishes.js [hunk 8]
```diff
@@ -69,13 +64,13 @@ class WishesAPI {
 
   /**
    * Get the top X pending wishes belonging to a category, sorted by timestamp
-   * @param {object} categoryId The category id
-   * @param {number} n The number of wishes to return
+   * @param {string} categoryId The category id
+   * @param {number} n The number of wishes to query
    * @throws {FirebaseError}
-   * @return {object} A firebase document of the top n pending wishes
+   * @return {array} A list of firebase document of the top n pending wishes
```
[the remark this answers is not in the document]
Comment (jinyingtan) [reply utils/api/wishes.js:70]: This return an array of wishes

File: utils/api/wishes.js [hunk 9]
```diff
@@ -1,48 +1,43 @@
...
+   * @param {string} description The wish description text
+   * @param {array} categories A list of categories id that the wish belongs to
+   * @throws {WishError}
    * @throws {FirebaseError}
-   * @return {boolean} true if the wish is created, false otherwise
+   * @return {object} A firebase document of the created wish
    */
   async create(title, description, categories) {
-    let categoriesInfo = [];
+    // time: 1051ms -> 760ms
```
Comment (jamessspanggg) [inline utils/api/wishes.js:20]: perhaps can remove the time improvements?

**Comment by jamessspanggg** [comment]
I think after this is merged, can proceed with changing all the wishes in the db to use wishId and use wishId in the front-end

**Comment by kohchihao** [comment]
Change already

**Comment by jinyingtan** [comment]
@jamessspanggg changed the `wishesId` to `wishId` on `TopWishes` and `WishesViewCategoryPage`

**Pull request merged**